# Report

Для анализа датасета Movielens создадим экземпляры классов Movies, Tags, Links, Ratings.

In [68]:
from movielens_analysis import Movies, Tags, Links, Ratings
import timeit

In [69]:
%%time
links = Links("data/links.txt", "data/film_info.txt", "data/list_of_fields.csv")
movies = Movies("../datasets/movies.csv")
tags = Tags("../datasets/tags.csv")
ratings = Ratings("../datasets/ratings.csv")

CPU times: total: 109 ms
Wall time: 511 ms


### Part 1: Movies

С помощью метода `dist_by_release()` класса `Movies` мы можем сделать вывод, что наибольшее количество фильмов **(311)** было выпущено в **2002** году, а наименьшее **(1)** - в **1922, 1919, 1921, 1915, 1917, 1902, 1903, 1908**. Также датасет содержит **выброс** - значение года 0.


In [ ]:
%%time
movies.dist_by_release()

CPU times: total: 0 ns
Wall time: 0 ns


{2002: 311,
 2006: 295,
 2001: 294,
 2007: 284,
 2000: 283,
 2009: 282,
 2003: 279,
 2004: 279,
 2014: 278,
 1996: 276,
 2015: 274,
 2005: 273,
 2008: 269,
 1999: 263,
 1997: 260,
 1995: 259,
 1998: 258,
 2011: 254,
 2010: 247,
 2013: 239,
 1994: 237,
 2012: 233,
 2016: 218,
 1993: 198,
 1992: 167,
 1988: 165,
 1987: 153,
 1990: 147,
 1991: 147,
 2017: 147,
 1989: 142,
 1986: 139,
 1985: 126,
 1984: 101,
 1981: 92,
 1980: 89,
 1982: 87,
 1983: 83,
 1979: 69,
 1977: 63,
 1973: 59,
 1978: 59,
 1965: 47,
 1971: 47,
 1974: 45,
 1976: 44,
 1964: 43,
 1967: 42,
 1968: 42,
 1975: 42,
 1966: 42,
 2018: 41,
 1962: 40,
 1972: 39,
 1963: 39,
 1959: 37,
 1960: 37,
 1955: 36,
 1969: 35,
 1961: 34,
 1970: 33,
 1957: 33,
 1958: 31,
 1953: 30,
 1956: 30,
 1940: 25,
 1949: 25,
 1954: 23,
 1942: 23,
 1939: 23,
 1946: 23,
 1951: 22,
 1950: 21,
 1947: 20,
 1948: 20,
 1941: 18,
 1936: 18,
 1945: 17,
 1937: 16,
 1952: 16,
 1944: 16,
 1938: 15,
 1931: 14,
 1935: 13,
 0: 13,
 1933: 12,
 1934: 11,
 1943: 10,
 

В тройку фильмов с самым большим количеством жанров, согласно методу `most_genres()`, вошли **"Rubber (2010)"** - 10 жанров, **Patlabor: The Movie (1989)** - 8 жанров, **"Mulan (1998)"** - 7 жанров. Датасет также содержит другие фильмы, у которых по 7 жанров, однако значения 8 и 10 не повторяются.

In [71]:
%%time
movies.most_genres(5)

CPU times: total: 0 ns
Wall time: 15.3 ms


{'Rubber (2010)': 10,
 'Patlabor: The Movie (Kidô keisatsu patorebâ: The Movie) (1989)': 8,
 'Mulan (1998)': 7,
 'Who Framed Roger Rabbit? (1988)': 7,
 'Osmosis Jones (2001)': 7}

С помощью метода `dist_by_genres()` мы можем оценить распределение фильмов по жанрам и отмечаем, что больше всего **(4361)** выпускается фильмов в жанре **Драма**, тогда как меньше всего **(87)** было выпущено фильмов жанра **Нуар**. Также датасет содержит позиции, отмеченные как **no genres listed** - их насчитывается **34**.

In [72]:
%%time
movies.dist_by_genres()

CPU times: total: 0 ns
Wall time: 10.4 ms


{'Drama': 4361,
 'Comedy': 3756,
 'Thriller': 1894,
 'Action': 1828,
 'Romance': 1596,
 'Adventure': 1263,
 'Crime': 1199,
 'Sci-Fi': 980,
 'Horror': 978,
 'Fantasy': 779,
 'Children': 664,
 'Animation': 611,
 'Mystery': 573,
 'Documentary': 440,
 'War': 382,
 'Musical': 334,
 'Western': 167,
 'IMAX': 158,
 'Film-Noir': 87,
 '(no genres listed)': 34}

Согласно методу `dist_by_genres()` выше, тройка самых поплярных жанров - это Драма, Комедия и Триллер. Посмотрим, есть ли фильмы, которые сочетают в себе все три жанра - используем метод `genre_filter()`. Да, такие действительно есть - немного от общего числа, но достаточно.

In [73]:
%%time
movies.genre_filter(['Drama', 'Comedy', 'Thriller'])

CPU times: total: 0 ns
Wall time: 3.54 ms


['Money Train (1995)',
 'To Die For (1995)',
 'Bad Boys (1995)',
 'Pulp Fiction (1994)',
 'Shallow Grave (1994)',
 'What Happened Was... (1994)',
 'Ghost (1990)',
 'Fargo (1996)',
 'Freeway (1996)',
 'Metro (1997)',
 'Suicide Kings (1997)',
 'Out of Sight (1998)',
 "Man Bites Dog (C'est arrivé près de chez vous) (1992)",
 'Nurse Betty (2000)',
 'Beautiful Creatures (2000)',
 'Shadows and Fog (1991)',
 'Osmosis Jones (2001)',
 'Stunt Man, The (1980)',
 'Another 48 Hrs. (1990)',
 'Cherish (2002)',
 'Wasabi (2001)',
 'Rules of Attraction, The (2002)',
 'Confessions of a Dangerous Mind (2002)',
 'Poolhall Junkies (2002)',
 'Party Monster (2003)',
 'Interstate 60 (2002)',
 'Last Boy Scout, The (1991)',
 'Ichi the Killer (Koroshiya 1) (2001)',
 'Dead or Alive: Final (2002)',
 '11:14 (2003)',
 'Seven-Per-Cent Solution, The (1976)',
 'Matador, The (2005)',
 "Pusher III: I'm the Angel of Death (2005)",
 'Host, The (Gwoemul) (2006)',
 'Hunting Party, The (2007)',
 'Southland Tales (2006)',
 'In 

Рассмотрим динамику частотности жанров по годам с помощью `genre_frequency_by_year`. Мы увидим, что Драма и Комедиz встречаются на лидирующих позициях чаще всего, иногда сменяя друг друга, однако иногда первые позиции занимают другие жанры - например, Анимация. Статистические такие случаи неважны, но могут сказаться на вычислениях.

In [74]:
%%time
movies.genre_frequency_by_year(3)

CPU times: total: 0 ns
Wall time: 15 ms


{1995: {'Drama': 123, 'Comedy': 91, 'Romance': 49},
 1994: {'Drama': 114, 'Comedy': 111, 'Romance': 41},
 1996: {'Drama': 128, 'Comedy': 108, 'Thriller': 61},
 1976: {'Drama': 21, 'Comedy': 16, 'Thriller': 15},
 1992: {'Drama': 78, 'Comedy': 69, 'Romance': 36},
 1967: {'Drama': 23, 'Comedy': 14, 'Action': 8},
 1993: {'Drama': 99, 'Comedy': 77, 'Action': 43},
 1964: {'Drama': 17, 'Comedy': 17, 'Romance': 7},
 1977: {'Drama': 27, 'Comedy': 21, 'Adventure': 12},
 1965: {'Drama': 24, 'Comedy': 17, 'Action': 9},
 1982: {'Comedy': 32, 'Drama': 29, 'Horror': 15},
 1990: {'Drama': 62, 'Comedy': 59, 'Action': 35},
 1991: {'Comedy': 69, 'Drama': 63, 'Romance': 31},
 1989: {'Comedy': 71, 'Drama': 60, 'Action': 33},
 1937: {'Drama': 12, 'Romance': 4, 'Musical': 3},
 1940: {'Drama': 14, 'Comedy': 11, 'Romance': 9},
 1969: {'Drama': 17, 'Comedy': 11, 'Adventure': 6},
 1981: {'Comedy': 32, 'Drama': 31, 'Action': 20},
 1973: {'Drama': 29, 'Crime': 17, 'Thriller': 14},
 1970: {'Drama': 18, 'Comedy': 10

Продолжим изучение самого популярного жанра - Драмы. Найдём пять жанров, с которыми создатели сочетают её чаще всего. Для этого используем метод `genre_correlation()`. Мы видим, что чаще всего Драму сочетают, как ни странно, с Комедией - более 1000 картин. Также в пятерку вошли Романтика, Триллер, Детектив и Экшен.

In [75]:
%%time
movies.genre_correlation('Drama', 5)

CPU times: total: 0 ns
Wall time: 5.09 ms


{'Drama': {'Comedy': 1013,
  'Romance': 934,
  'Thriller': 832,
  'Crime': 637,
  'Action': 529}}

Узнаем, как часто фильмы в этом жанре выходили в разные годы, с помощью `definite_genre_frequency_by_year()`. Мы видим, что больше всего фильмов жанра Драма выходило в 2002 году, меньше всего - в первые 2 десятилетия ХХ века, что, однако, объясняется состоянием кинематографа того периода.

In [76]:
%%time
movies.definite_genre_frequency_by_year('Drama')

CPU times: total: 0 ns
Wall time: 2.31 ms


{'Drama': {2002: 150,
  2001: 148,
  2006: 146,
  2007: 142,
  2005: 140,
  2000: 138,
  2009: 131,
  2003: 129,
  1996: 128,
  2004: 128,
  1997: 125,
  1999: 124,
  1995: 123,
  1998: 122,
  2014: 117,
  1994: 114,
  2008: 113,
  2011: 110,
  2010: 102,
  2015: 101,
  1993: 99,
  2013: 98,
  2012: 92,
  1992: 78,
  2016: 69,
  1988: 67,
  1991: 63,
  1990: 62,
  1989: 60,
  1987: 56,
  1986: 55,
  1985: 45,
  2017: 45,
  1984: 35,
  1979: 32,
  1980: 32,
  1981: 31,
  1973: 29,
  1982: 29,
  1983: 29,
  1977: 27,
  1965: 24,
  1962: 24,
  1967: 23,
  1972: 22,
  1976: 21,
  1955: 21,
  1971: 20,
  1974: 20,
  1960: 19,
  1975: 19,
  1961: 18,
  1957: 18,
  1956: 18,
  1970: 18,
  1966: 18,
  1964: 17,
  1953: 17,
  1968: 17,
  1978: 17,
  1969: 17,
  1959: 15,
  1939: 15,
  1954: 15,
  1963: 15,
  1948: 15,
  1949: 15,
  1940: 14,
  1946: 14,
  1942: 13,
  1937: 12,
  1947: 12,
  1951: 12,
  1958: 11,
  1944: 11,
  1950: 10,
  1941: 9,
  1938: 9,
  1952: 9,
  1931: 9,
  2018: 9,
  19

Из всего вышеперечисленного можно сделать вывод, что Драма и Комедия являются самыми популярными жанрами кинематографа, при этом они также чаще всего сочетаются между собой. Всплеск жанра пришёлся на первые годы ХХI столетия - именно тогда снималось больше всего фильмов в жанре Драма. Это, однако, коррелирует со всплеском релизов в эти годы. Можно сказать, что в целом жанровое распределение по годам стабильно.

### Part 2: Tags

Теперь проанализируем тэги в нашем датасете. Для начала оценим максимально возможную длину тэгов. Для этого воспользуемся методом `most_words()` класса `Tags` и выведем 10 тэгов с максимальным количеством слов. Как мы видим, максимальное число - **16** - сильно выбивается и, если оно будет единичным, может считаться выбросом.

In [77]:
%%time
tags.most_words(10)

CPU times: total: 0 ns
Wall time: 0 ns


{'Something For Everyone In This One... Saw It Without And Plan On Seeing It With Kids!': 16,
 'The Catholic Church Is The Most Corrupt Organization In History': 10,
 'Villain Nonexistent Or Not Needed For Good Story': 8,
 '06 Oscar Nominated Best Movie - Animation': 7,
 'It Was Melodramatic And Kind Of Dumb': 7,
 'Oscar (Best Music - Original Score)': 6,
 'Stop Using Useless Characters For Filler': 6,
 'Oscar (Best Effects - Visual Effects)': 6,
 'Everything You Want Is Here': 5,
 'Based On A True Story': 5}

После этого логично узнать самые длинные тэги, измеряя длину не в словах, а в символах. Для этого воспользуемся методом `longest()`. Выводим 10 самых длинных тэгов и видим, что списки совпадают лишь частично. 

In [78]:
%%time
tags.longest(10)

CPU times: total: 0 ns
Wall time: 626 μs


['Something For Everyone In This One... Saw It Without And Plan On Seeing It With Kids!',
 'The Catholic Church Is The Most Corrupt Organization In History',
 'Villain Nonexistent Or Not Needed For Good Story',
 'R:Disturbing Violent Content Including Rape',
 '06 Oscar Nominated Best Movie - Animation',
 'Stop Using Useless Characters For Filler',
 'Academy Award (Best Supporting Actress)',
 'Oscar (Best Effects - Visual Effects)',
 'Audience Intelligence Underestimated',
 'It Was Melodramatic And Kind Of Dumb']

Учитывая выводы выше, сопоставим два списка и выведем пересечения с помощью метода `most_words_and_longest`. Получим **7** наименований.

In [79]:
%%time
tags.most_words_and_longest(10)

CPU times: total: 15.6 ms
Wall time: 3.03 ms


['06 Oscar Nominated Best Movie - Animation',
 'It Was Melodramatic And Kind Of Dumb',
 'Oscar (Best Effects - Visual Effects)',
 'Something For Everyone In This One... Saw It Without And Plan On Seeing It With Kids!',
 'Stop Using Useless Characters For Filler',
 'The Catholic Church Is The Most Corrupt Organization In History',
 'Villain Nonexistent Or Not Needed For Good Story']

Теперь рассмотрим самые популярные тэги за всё время - воспользуемся методом `most_popular()`. Как мы видим, десятку самых популярных тэгов возглавляет тэг **"In Netflix Queue"**. Он лидирует с большим отрывом - **131 упоминание**.

In [80]:
%%time
tags.most_popular(10)

CPU times: total: 0 ns
Wall time: 0 ns


{'In Netflix Queue': 131,
 'Atmospheric': 41,
 'Funny': 24,
 'Superhero': 24,
 'Surreal': 24,
 'Thought-Provoking': 24,
 'Sci-Fi': 23,
 'Disney': 23,
 'Quirky': 22,
 'Religion': 22}

Узнаем, является ли Netflix, в таком случае, популярным упоминанием в других тэгах. Вывод `tags_with()` показывает, что это не так. Тем не менее, он обнаружил три тэга с упоминанием хостинга. 

In [81]:
%%time
tags.tags_with('Netflix')

CPU times: total: 0 ns
Wall time: 0 ns


['In Netflix Queue', 'No Dvd At Netflix', 'Not Available From Netflix']

Рассмотрим, является ли использование тэгов стабильным по годам. Воспользуемся методом `tags_for_years()`. Видим, что резкий всплеск пришёлся на **2006 год**, после чего количество тэгов резко упало и показывает растущую тенденуию только **с 2015 года**. 

In [82]:
%%time
tags.tags_for_years()

CPU times: total: 0 ns
Wall time: 3.51 ms


{2006: 1533,
 2018: 844,
 2016: 355,
 2017: 329,
 2015: 191,
 2009: 166,
 2010: 133,
 2012: 47,
 2007: 46,
 2011: 13,
 2013: 10,
 2008: 9,
 2014: 7}

Установим самые популярные тэги в 2006 году и сравним их с теми, которые использовались в 2014 - в год с самым малым использованием тэгов. Воспользуемся методом `most_popular_by_year()`. Пересечений, однако, не находим. Однако видим, что именно в 2006 году был использован тэг **"In Netflix Queue"**, который до сих пор удерживает планку самого популярного. До и после он не использовался (131 упоминание всего, 131 упоминание в 2006 году). Это может иметь значение для статистики, особенно учитывая, что именно в 2006-2007 году произошла трансформация Neetflix-a в стриминговый сервис.

In [83]:
%%time
tags.most_popular_by_year(2006, 7)

CPU times: total: 0 ns
Wall time: 0 ns


{'In Netflix Queue': 131,
 'Disney': 21,
 'Religion': 20,
 'Superhero': 14,
 'Politics': 14,
 'High School': 13,
 'Journalism': 13}

In [84]:
%%time
tags.most_popular_by_year(2014, 7)

CPU times: total: 0 ns
Wall time: 0 ns


{'Reciprocal Spectator': 2,
 'Romantic Comedy': 1,
 'Wedding': 1,
 'Painter': 1,
 'Bloody': 1,
 'Creature Feature': 1}

Проанализируем другие тэги, востребованные в 2006 году - посмотрим, использовались ли они в другие годы. Для этого используем метод `count_definite_by_year()`. Видим, что все тэги, за исключением **"In Netflix Queue"** и **"Journalism"** были востребованы и после 2006 года (более ранними данными не располагаем). Среди них тэги **"Disney"**, **"Religion"** и **"Superhero"** остаются среди самых популярных на 2018 год (более поздними данными не располагаем).

In [85]:
%%time
tags.count_definite_by_year(['In Netflix Queue', 'Disney', 'Religion', 'Superhero', 'Politics', 'High School', 'Journalism'])

CPU times: total: 0 ns
Wall time: 0 ns


{'In Netflix Queue': {2006: 131},
 'Disney': {2018: 1, 2016: 1, 2006: 21},
 'Religion': {2006: 20, 2009: 1, 2010: 1},
 'Superhero': {2018: 7, 2006: 14, 2012: 1, 2013: 1, 2015: 1},
 'Politics': {2018: 1, 2006: 14, 2015: 4},
 'High School': {2015: 1, 2012: 1, 2006: 13, 2009: 1},
 'Journalism': {2006: 13}}

Теперь посмотрим топ тэгов по количествую уникальных фильмов, которые были ими отмечены. Воспользуемся методом `tag_count_by_distinct_movies()`.

In [86]:
tags.tag_count_by_distinct_movies(10)

{'In Netflix Queue': 131,
 'Atmospheric': 37,
 'Funny': 22,
 'Quirky': 22,
 'Disney': 22,
 'Superhero': 22,
 'Surreal': 22,
 'Religion': 22,
 'Sci-Fi': 19,
 'Comedy': 19}

Наконец, для более чёткого понимания распределения тэгов будет полезно узнать, сколько уникальных пользователей их использовали. Выведем 10 тэгов, которые использует наибольшее число людей - воспользуемся методом `tag_count_by_distinct_users()`.

In [87]:
%%time
tags.tag_count_by_distinct_users(10)

CPU times: total: 0 ns
Wall time: 7.62 ms


{'Funny': 10,
 'Sci-Fi': 10,
 'Atmospheric': 10,
 'Dark Comedy': 9,
 'Comedy': 9,
 'Music': 8,
 'Mindfuck': 8,
 'Suspense': 8,
 'Twist Ending': 7,
 'Animation': 7}

Мы видим, что хотя некоторые из самых популярных тэгов попали в список выше, самого популярного тэга здесь нет. Это наводит на мысль о том, что стоит проверить, сколько пользователей в действительности пользовалось этим тэгом. Так и есть - метод `distinct_tag_count_by_distinct_users()` показывает, что пользовался этим тэгом всего один человек.

In [88]:
%%time
tags.distinct_tag_count_by_distinct_users('In Netflix Queue')

CPU times: total: 0 ns
Wall time: 5.02 ms


1

Итак, анализ данных по тэгам дал противоречивые результаты. С одной стороны, мы установили пик использования тэгов и вывели самые популярные тэги за всё время. С другой стороны - выяснили, что самый популярный тэг использовался лишь одним человеком и только в одном году, а значит, вряд ли другие люди станут искать фильмы по этому тэгу. Несколько больше результатов даст сопоставление самых популярных тэгов с самыми распространенными среди уникальных пользователей. Мы видим следующие пересечения:
 - **Atmospheric**
 - **Funny**

 Эти тэги не только являются самыми популярными, но и используются наибольшим количеством пользователей. 

In [89]:
%%time
tags.most_popular(10)

CPU times: total: 0 ns
Wall time: 0 ns


{'In Netflix Queue': 131,
 'Atmospheric': 41,
 'Funny': 24,
 'Superhero': 24,
 'Surreal': 24,
 'Thought-Provoking': 24,
 'Sci-Fi': 23,
 'Disney': 23,
 'Quirky': 22,
 'Religion': 22}

In [90]:
%%time
tags.tag_count_by_distinct_users(10)

CPU times: total: 0 ns
Wall time: 8.08 ms


{'Funny': 10,
 'Sci-Fi': 10,
 'Atmospheric': 10,
 'Dark Comedy': 9,
 'Comedy': 9,
 'Music': 8,
 'Mindfuck': 8,
 'Suspense': 8,
 'Twist Ending': 7,
 'Animation': 7}

Эти же тэги входят в топ по количеству уникальных фильмов. Так что можно сделать вывод, что именно на них следует опираться, если нужно, например, предсказать поведение пользователей онлайн-кинотеатра. 

### Part 3: Ratings

Теперь проанализируем информацию о рейтингах, доступную нам из датасета. Для этого на основе класса `Ratings` создадим две дополнительных сущности - `Ratings.Movies` и `Ratings.Users`.

In [91]:
%%time
mv = ratings.Movies(ratings)
us = ratings.Users(ratings)

CPU times: total: 0 ns
Wall time: 10 ms


##### Ratings.Movies

Оценим рейтинги, связанные с фильмами. Для начала определим количество оценок в год с помощью метода `dist_by_year()` класса `Ratings.Movies`. Мы видим, что лидером по количеству оценок стал **2000 год**. Также достаточно большое количество оценок было в 1996, 2007, 2017 году. Между этими точками наблюдаются периоды спада, однако можно говорить о достаточно равномерном распределении рейтингов и о стабильно высоких показателях с 2015 по 2018 год.

In [92]:
%%time
mv.dist_by_year()

CPU times: total: 0 ns
Wall time: 11.2 ms


{1996: 6040,
 1997: 1916,
 1998: 507,
 1999: 2439,
 2000: 10061,
 2001: 3922,
 2002: 3478,
 2003: 4014,
 2004: 3279,
 2005: 5813,
 2006: 4059,
 2007: 7114,
 2008: 4351,
 2009: 4158,
 2010: 2300,
 2011: 1690,
 2012: 4657,
 2013: 1664,
 2014: 1439,
 2015: 6616,
 2016: 6702,
 2017: 8199,
 2018: 6418}

Теперь посмотрим на распределение самих оценок с помощью метода `dist_by_rating()`. Видим, что **минимальная оценка (0.5)** была поставлена 1370 раза, а **максимальная (5.0)** - 13211 раз. **Наиболее частые оценки** - 4.0 (26818), 3.0 (20047), 5.0 (13211). Обратим внимание, что, во-первых, пользователи склонны ставить оценки без дробной части, а во-вторых, они чаще ставят положительные оценки. Это, однако, не может быть свидетельством среднего качества фильмов, но может кое-что сказать о самих зрителях: вероятно, они чаще оценивают фильм, если он им понравился, если нет - оставляют без оценки. 

In [93]:
%%time
mv.dist_by_rating()

CPU times: total: 0 ns
Wall time: 19 ms


{0.5: 1370,
 1.0: 2811,
 1.5: 1791,
 2.0: 7551,
 2.5: 5550,
 3.0: 20047,
 3.5: 13136,
 4.0: 26818,
 4.5: 8551,
 5.0: 13211}

Теперь выведем топ 10 самых оцениваемых фильмов, воспользовавшись методом `top_by_num_of_ratings()`. Видим, что **максимальное количество оценок** для фильма составляет **329**, причем всё это - фольмы, выпущенные до 2000 года, когда происходит резкий рост количества оценок. Это может иметь значение для статистики.

In [94]:
%%time
mv.top_by_num_of_ratings(10)

CPU times: total: 0 ns
Wall time: 27.2 ms


{'Forrest Gump (1994)': 329,
 'Shawshank Redemption, The (1994)': 317,
 'Pulp Fiction (1994)': 307,
 'Silence of the Lambs, The (1991)': 279,
 'Matrix, The (1999)': 278,
 'Star Wars: Episode IV - A New Hope (1977)': 251,
 'Jurassic Park (1993)': 238,
 'Braveheart (1995)': 237,
 'Terminator 2: Judgment Day (1991)': 224,
 "Schindler's List (1993)": 220}

Выведем топ 10 фильмов по средней оценке пользователей. Если наша теория верна, то все они будут оценены на 5.0 - таких оценок должно быть много. Как видно, данные, собранные методом `top_by_ratings()`, это подтверждают.

In [95]:
%%time
mv.top_by_ratings(10)

CPU times: total: 15.6 ms
Wall time: 42 ms


{'The Jinx: The Life and Deaths of Robert Durst (2015)': 5.0,
 'Galaxy of Terror (Quest) (1981)': 5.0,
 'Alien Contamination (1980)': 5.0,
 "I'm the One That I Want (2000)": 5.0,
 'Lesson Faust (1994)': 5.0,
 'Assignment, The (1997)': 5.0,
 'Mephisto (1981)': 5.0,
 'Black Mirror': 5.0,
 'Dylan Moran: Monster (2004)': 5.0,
 'Bill Hicks: Revelations (1993)': 5.0}

Также интересно было бы узнать возможную дисперсию оценок. Для этого вопользуемс `top_controversial()` и выведем том 10 фильмов с наиболее противоречивыми оценками. Как мы видим, датасет действительно содержит фильмы с широким диапазоном дисперсии оценок - максимальный разброс составил **5.06**.

In [96]:
%%time
mv.top_controversial(10)

CPU times: total: 0 ns
Wall time: 96 ms


{"id: 32892, title: Ivan's Childhood (a.k.a. My Name is Ivan) (Ivanovo detstvo) (1962)": 5.06,
 'id: 2068, title: Fanny and Alexander (Fanny och Alexander) (1982)': 5.06,
 'id: 70946, title: Troll 2 (1990)': 4.5,
 'id: 484, title: Lassie (1994)': 4.0,
 'id: 3223, title: Zed & Two Noughts, A (1985)': 4.0,
 'id: 7564, title: Kwaidan (Kaidan) (1964)': 4.0,
 'id: 84847, title: Emma (2009)': 4.0,
 'id: 26171, title: Play Time (a.k.a. Playtime) (1967)': 3.72,
 'id: 74754, title: Room, The (2003)': 3.56,
 'id: 2488, title: Peeping Tom (1960)': 3.5}

Подобная дисперсия вызывает интерес. Выведем дисперсию оценок отдельно для фильмов **"BamIvan's Childhood (a.k.a. My Name is Ivan) (Ivanovo detstvo) (1962)"** и **"Fanny and Alexander (Fanny och Alexander) (1982)"** чтобы узнать, есть ли в данных выбросы. Воспользуемся методом `rating_histogram()`. Функция получает ID фильма. ID нужных нам фильмов - 32892, 2068. Везде видим одинаковую картину: рейтинг складывается из двух оценок, максимальной и минимальной.

In [97]:
%%time
mv.rating_histogram(32892)

CPU times: total: 0 ns
Wall time: 7.59 ms


{0.5: 1, 5.0: 1}

In [98]:
%%time
mv.rating_histogram(2068)

CPU times: total: 0 ns
Wall time: 1.5 ms


{0.5: 1, 5.0: 1}

Теперь рассмотрим один из фильмов с наибольшим количеством оценок - Forrest Gump (1994) (ID=356).  Узнаем, в какие годы эти оценки выставлляись и какова была средняя оценка для каждого года, с помощью метода `rating_trend()`. На выводе видим, что фильм был встречен хорошими оценками, а в 1998 году средний рейтинг и вовсе составил 5.0. До 2018 года это значение не опускалось ниже 3.12 - самый низкий рейтинг, который был присвоен ему в 2014 году.

In [99]:
%%time
mv.rating_trend(50)

CPU times: total: 0 ns
Wall time: 0 ns


{1996: 4.03,
 1997: 4.2,
 1998: 5.0,
 1999: 4.38,
 2000: 4.35,
 2001: 4.67,
 2002: 4.75,
 2003: 4.25,
 2004: 4.5,
 2005: 4.31,
 2006: 3.94,
 2007: 4.55,
 2008: 4.35,
 2009: 4.64,
 2010: 4.75,
 2011: 3.67,
 2012: 4.5,
 2013: 4.33,
 2014: 3.12,
 2015: 4.23,
 2016: 4.42,
 2017: 4.27,
 2018: 3.69}

Таким образом в контексте рейтингов фильмов мы можем сделать вывод об общей стабильности (стационарности) количества оценок, при этом пользователи активнее оценивают фильмы, которые им понравились, и выставляют больше положительных оценок (>=3). Общая активность невысока (максимум 329 оценок на фильм), однако в последние годы датасета количество оценок стабильно высокое. В то же время, в датасете присутствуют фильмы с большой дисперсией оценок, что может сказаться на качестве анализа. 

Однако, чтобы составить полную картину, необходимо проанализировать рейтинги в контексте пользователей. 

##### Ratings.Users

Оценим поведение зрителей. Для начала узнаем распределение зрителей (ID) по количеству оценок. Воспользуемся методом `dist_by_num_of_ratings()` класса `Users`. Видим, что максимум один пользователь за весь период выставлял 2698 оценок. Наименьшее количество составило 20 - столько оценок поставили 14 человек.

In [100]:
%%time
us.dist_by_num_of_ratings()

CPU times: total: 0 ns
Wall time: 28.3 ms


{20: 14,
 21: 15,
 22: 14,
 23: 13,
 24: 7,
 25: 9,
 26: 13,
 27: 8,
 28: 8,
 29: 8,
 30: 3,
 31: 7,
 32: 7,
 33: 10,
 34: 9,
 35: 11,
 36: 9,
 37: 6,
 38: 7,
 39: 4,
 40: 6,
 41: 4,
 42: 5,
 43: 3,
 44: 3,
 45: 6,
 46: 5,
 47: 4,
 48: 7,
 50: 7,
 51: 5,
 52: 2,
 53: 4,
 54: 3,
 55: 2,
 56: 14,
 57: 4,
 58: 5,
 59: 3,
 60: 2,
 61: 5,
 62: 2,
 63: 1,
 64: 5,
 65: 4,
 66: 2,
 67: 2,
 68: 1,
 69: 5,
 70: 2,
 71: 1,
 72: 2,
 73: 2,
 74: 2,
 75: 3,
 76: 3,
 77: 3,
 78: 3,
 80: 1,
 81: 2,
 82: 2,
 83: 3,
 84: 3,
 85: 1,
 86: 2,
 87: 3,
 88: 2,
 89: 2,
 90: 1,
 92: 1,
 93: 4,
 94: 4,
 95: 1,
 97: 3,
 98: 3,
 100: 3,
 101: 1,
 102: 2,
 103: 3,
 105: 1,
 106: 2,
 107: 2,
 108: 1,
 109: 1,
 110: 2,
 111: 2,
 112: 4,
 113: 3,
 114: 2,
 115: 2,
 118: 2,
 119: 2,
 120: 1,
 121: 2,
 122: 1,
 123: 2,
 126: 1,
 127: 3,
 128: 4,
 129: 2,
 130: 1,
 131: 2,
 133: 1,
 134: 1,
 135: 3,
 138: 3,
 139: 1,
 140: 4,
 141: 4,
 148: 1,
 150: 3,
 152: 4,
 154: 1,
 155: 3,
 156: 1,
 157: 1,
 161: 1,
 163: 1,
 164:

Рассмотрим распределение по оценкам с помощью метода `dist_by_rating()`. Видим, что для большинства пользователей характерны средние оценки от 3.5 до 5 с явным преобладанием значений между 4.0 и 5.0.

In [101]:
%%time
us.dist_by_rating()

CPU times: total: 0 ns
Wall time: 27.5 ms


{1.27: 1,
 2.14: 1,
 2.15: 1,
 2.22: 1,
 2.25: 1,
 2.34: 1,
 2.36: 1,
 2.39: 1,
 2.43: 1,
 2.44: 1,
 2.57: 3,
 2.6: 1,
 2.61: 2,
 2.62: 2,
 2.63: 1,
 2.64: 3,
 2.67: 2,
 2.71: 2,
 2.72: 1,
 2.73: 1,
 2.75: 2,
 2.76: 1,
 2.77: 1,
 2.78: 1,
 2.81: 1,
 2.82: 1,
 2.84: 2,
 2.85: 1,
 2.86: 2,
 2.87: 1,
 2.88: 1,
 2.9: 1,
 2.94: 4,
 2.96: 1,
 2.99: 1,
 3.0: 3,
 3.02: 2,
 3.03: 1,
 3.04: 2,
 3.05: 3,
 3.07: 1,
 3.08: 1,
 3.09: 1,
 3.1: 2,
 3.11: 2,
 3.12: 2,
 3.13: 1,
 3.14: 4,
 3.15: 1,
 3.16: 2,
 3.17: 4,
 3.18: 2,
 3.19: 2,
 3.2: 5,
 3.21: 3,
 3.22: 3,
 3.23: 4,
 3.24: 3,
 3.25: 5,
 3.26: 6,
 3.27: 5,
 3.28: 3,
 3.29: 5,
 3.3: 4,
 3.31: 4,
 3.32: 6,
 3.34: 4,
 3.35: 3,
 3.36: 7,
 3.37: 3,
 3.38: 4,
 3.39: 7,
 3.4: 7,
 3.41: 6,
 3.42: 6,
 3.43: 2,
 3.44: 5,
 3.45: 1,
 3.46: 4,
 3.47: 2,
 3.48: 5,
 3.49: 3,
 3.5: 4,
 3.51: 6,
 3.52: 2,
 3.53: 6,
 3.54: 10,
 3.55: 2,
 3.56: 5,
 3.57: 8,
 3.58: 2,
 3.59: 4,
 3.6: 4,
 3.61: 1,
 3.62: 4,
 3.63: 5,
 3.64: 3,
 3.65: 9,
 3.66: 4,
 3.67: 5,
 3.68: 5

Также в датасете присутствуют пользователи с большим показателем дисперсии оценок. Из вывода метода `top_controversial()` наблюдаем, что наибольшая дисперсия составила 4.26 у пользователя с ID=3. 

In [102]:
%%time
us.top_controversial(10)

CPU times: total: 0 ns
Wall time: 39.7 ms


{3: 4.26,
 461: 3.1,
 55: 3.09,
 259: 2.94,
 329: 2.92,
 502: 2.76,
 175: 2.75,
 598: 2.7,
 393: 2.61,
 138: 2.44}

Для статистики важна активность зрителей. Выведем ID десяти самых активных пользователей методом `top_by_num_of_ratings()`. Видим, что в десятку вошли зрители, поставившие более 1000 оценок.

In [103]:
%%time
us.top_by_num_of_ratings(10)

CPU times: total: 0 ns
Wall time: 14.9 ms


{414: 2698,
 599: 2478,
 474: 2108,
 448: 1864,
 274: 1346,
 610: 1302,
 68: 1260,
 380: 1218,
 606: 1115,
 288: 1055}

С помощью метода `top_generous()` также установим 10 пользователей, склонных давать наилучшие оценки фильмам. Эта категория пользователей может в перспективе составить таргетную группу. Как мы видим, средняя оценка, выставляемая этими пользователям, составляет более 4.5.

In [104]:
%%time
us.top_generous(10)

CPU times: total: 15.6 ms
Wall time: 10.2 ms


{53: 5.0,
 251: 4.87,
 515: 4.85,
 25: 4.81,
 30: 4.74,
 523: 4.69,
 348: 4.67,
 171: 4.63,
 452: 4.56,
 43: 4.55}

### Part 4: Links

Класс Links позволяет нам провести анализ данных фильмов, представленных в links.csv.
 
 С помощью функции `top_directors()` мы устанавливаем топ-10 режиссеров по количеству отснятых фильмов.
  Таким образом, мы усиановили, что `Альфред Хичкок`, отснял больше всего фильмов `(12 шт)` в сравнении с остальными претендентами, указанными в списке. 

In [105]:
%%time 
links.top_directors(10)

CPU times: total: 0 ns
Wall time: 0 ns


{'Forest Whitaker': 1,
 'John Lasseter': 1,
 'Peter Hyams': 1,
 'Sydney Pollack': 1,
 'Oliver Stone': 1,
 'Joe Johnston': 1,
 'Michael Mann': 1,
 'Howard Deutch': 1,
 'Martin Campbell': 1,
 'Charles Shyer': 1}

С помощью функции `most_expensive()` мы устанавливаем топ-3 самых дорогих фильмов.
 

In [106]:
%%time 
links.most_expensive(3)

CPU times: total: 0 ns
Wall time: 0 ns


{'Jumanji': 65000000, 'The American President': 62000000, 'Heat': 60000000}

С помощью функции `the_cheapest()` мы устанавливаем топ-3 самых дешевых фильмов.

In [107]:
%%time 
links.the_cheapest(3)

CPU times: total: 0 ns
Wall time: 3.64 ms


{'Waiting to Exhale': 16000000,
 'Grumpier Old Men': 25000000,
 'Toy Story': 30000000}

С помощью функции `most_profitable()` мы устанавливаем топ-3 самых прибыльных фильмов.

In [108]:
%%time 
links.most_profitable(3)

CPU times: total: 0 ns
Wall time: 0 ns


{'Toy Story': 364436586, 'GoldenEye': 292194034, 'Jumanji': 197821940}

С помощью функции `less_profitable()` мы устанавливаем топ-3 самых неприбыльных фильмов.

In [109]:
%%time 
links.less_profitable(3)

CPU times: total: 0 ns
Wall time: 0 ns


{'Nixon': -30318235,
 'Dracula: Dead and Loving It': -19227856,
 'Sabrina': -4303041}

С помощью функции `longest()` мы устанавливаем топ-3 самых длинных фильмов. Длина измеряется в минутах.

In [110]:
%%time 
links.longest(3)

CPU times: total: 0 ns
Wall time: 0 ns


{'Nixon': 192, 'Heat': 170, 'GoldenEye': 130}

С помощью функции `shortest()` мы устанавливаем топ-3 самых коротких фильмов. Длина измеряется в минутах.

In [111]:
%%time 
links.shortest(3)

CPU times: total: 0 ns
Wall time: 0 ns


{'Balto': 78, 'Toy Story': 81, 'Dracula: Dead and Loving It': 88}

С помощью функции `top_cost_per_minute()` мы устанавливаем топ-3 фильмов с самой высокой стоимостью за минуту.

In [112]:
%%time 
links.top_cost_per_minute(3)

CPU times: total: 0 ns
Wall time: 0 ns


{'Jumanji': 625000.0,
 'The American President': 543859.65,
 'GoldenEye': 461538.46}

С помощью функции `less_cost_per_minute()` мы устанавливаем топ-3 фильмов с самой низкой стоимостью за минуту.

In [113]:
%%time 
links.less_cost_per_minute(3)

CPU times: total: 0 ns
Wall time: 0 ns


{'Waiting to Exhale': 129032.26,
 'Nixon': 229166.67,
 'Grumpier Old Men': 247524.75}

С помощью функции `top_longest_title()` мы устанавливаем топ-3 фильмов с самым длинным названием.

In [114]:
%%time 
links.top_longest_title(3)

CPU times: total: 0 ns
Wall time: 0 ns


['Father of the Bride Part II',
 'Dracula: Dead and Loving It',
 'The American President']

С помощью функции `top_shortest_title()` мы устанавливаем топ-3 фильмов с самым коротким названием.

In [115]:
%%time 
links.top_shortest_title(3)

CPU times: total: 0 ns
Wall time: 0 ns


['Heat', 'Nixon', 'Balto']